File: bug23_returning_wrong_value/solved_main.java
import java.nio.ByteBuffer;
import java.util.Base64;

public class solved_main {
    static long mix(long x) {
        x = (x ^ (x >>> 30)) * 0xBF58476D1CE4E5B9L;
        x = (x ^ (x >>> 27)) * 0x94D049BB133111EBL;
        x ^= x >>> 31;
        return x;   // FIX: was "return 0"; must return computed x
    }

    public static void main(String[] args) {
        long seed = 0x0102030405060708L;
        long acc = seed;
        for (long i=0; i<10; i++)
            acc = mix(acc ^ (i * 0x9E3779B97F4A7C15L));
        ByteBuffer buf = ByteBuffer.allocate(8);
        buf.putLong(acc);
        System.out.println(Base64.getEncoder().encodeToString(buf.array()));
    }
}

File: bug23_returning_wrong_value/solved_main.cpp
#include <iostream>
#include <cstdint>
#include <string>

std::string base64_encode(uint64_t val) {
    const std::string b="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    uint8_t bytes[8];
    for(int i=7;i>=0;i--){bytes[i]=val&0xFF;val>>=8;}
    std::string out;
    out+=b[bytes[0]>>2];out+=b[((bytes[0]&3)<<4)|(bytes[1]>>4)];
    out+=b[((bytes[1]&15)<<2)|(bytes[2]>>6)];out+=b[bytes[2]&63];
    out+=b[bytes[3]>>2];out+=b[((bytes[3]&3)<<4)|(bytes[4]>>4)];
    out+=b[((bytes[4]&15)<<2)|(bytes[5]>>6)];out+=b[bytes[5]&63];
    out+=b[bytes[6]>>2];out+=b[((bytes[6]&3)<<4)|(bytes[7]>>4)];
    out+=b[((bytes[7]&15)<<2)];out+="=";
    return out;
}

uint64_t mix(uint64_t x) {
    x=(x^(x>>30))*0xBF58476D1CE4E5B9ULL;
    x=(x^(x>>27))*0x94D049BB133111EBULL;
    x^=x>>31;
    return x;   // FIX: was "return 0"; must return computed x
}

int main() {
    uint64_t seed=0x0102030405060708ULL, acc=seed;
    for(uint64_t i=0;i<10;i++)
        acc=mix(acc^(i*0x9E3779B97F4A7C15ULL));
    std::cout<<base64_encode(acc)<<std::endl;
    return 0;
}
